### PR TITLE
support qwen3 tts streaming output

### PR DESCRIPTION
**Pull request by gerayking**
## Purpose
Add streaming output support for Qwen3-TTS model to enable real-time audio generation with lower latency. #938 

## Key Changes

1. **Streaming Generation Interface** (`modeling_qwen3_tts.py`):
   - Added `StreamingChunkOutput` dataclass for streaming chunk output
   - Added `AsyncDecodingPipeline` class for asynchronous audio decoding in background thread
   - Implemented `generate_streaming_iter()` in `Qwen3TTSTalkerForConditionalGeneration` for token-level streaming

2. **High-Level Streaming API** (`qwen3_tts.py`):
   - Added `forward_streaming()` method in `Qwen3TTSModelForGeneration`
   - Added streaming variants: `generate_custom_voice_streaming()`, `generate_voice_design_streaming()`, `generate_voice_clone_streaming()`
   - Implemented streaming state management for request handling

3. **Examples and Tests**:
   - Added streaming test cases in `examples/offline_inference/qwen3_tts/end2end.py`
   - Added `test_streaming_vs_nonstreaming.py` to verify streaming and non-streaming outputs produce identical tokens
## Test Plan

```bash
# Run streaming vs non-streaming consistency test
python tests/model_executor/models/qwen3_tts/test_streaming_vs_nonstreaming.py

# Run end2end streaming example
python examples/offline_inference/qwen3_tts/end2end.py --streaming
```

## Test Result

```
============================================================
COMPARISON RESULTS (TOKEN_ID LEVEL):
============================================================
Non-streaming tokens (raw): 88
Streaming tokens (raw):     89
EOS token ID: 4198
Non-streaming tokens (no EOS): 88
Streaming tokens (no EOS):     88
✓ Token counts match (excluding EOS)!
✓ All 88 tokens are IDENTICAL across all 16 codebooks!

============================================================
TEST PASSED!
============================================================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts.py [hunk 1]
```diff
@@ -150,6 +182,155 @@ def forward(
         # Convert result to OmniOutput format
         return self.make_omni_output(result, **kwargs)
 
+    def forward_streaming(
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts.py:184]: does every tts model need this function?

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts.py [hunk 2]
```diff
@@ -1086,3 +1267,279 @@ def get_supported_languages(self) -> list[str] | None:
         if supported is None:
             return None
         return sorted(supported)
+
+    # ==================== Streaming Generation Methods ====================
+
+    @torch.inference_mode()
+    def generate_custom_voice_streaming(
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/qwen3_tts.py:1279]: same question

File: vllm_omni/model_executor/models/qwen3_tts/modeling_qwen3_tts.py [hunk 3]
```diff
@@ -1324,6 +1328,167 @@ class Qwen3TTSTalkerOutputWithPast(ModelOutput):
     tts_pad_embed: torch.FloatTensor | None = None
 
 
+@dataclass
+class StreamingChunkOutput:
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/modeling_qwen3_tts.py]: for StreamingChunkOutput& AsyncDecodingPipeline abstraction, could they be used in other tts models?

**Review by hsliuustc0106 — COMMENTED** [review]
could this work with online serving?

**Comment by hsliuustc0106** [comment]
I think we need to discuss a general abstraction of streaming output such that every tts model can inherit from it. @Gaohan123 @linyueqian @amy-why-3459

File: tests/model_executor/models/qwen3_tts/test_streaming_vs_nonstreaming.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+    """
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    print(f"Using device: {device}")
+
+    # Load model
+    print(f"Loading model from {MODEL_PATH}...")
+    model = Qwen3TTSModel.from_pretrained(
+        MODEL_PATH,
+        torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32,
+        device_map=device,
```
[the remark this answers is not in the document]
Comment (marksverdhei) [reply tests/model_executor/models/qwen3_tts/test_streaming_vs_nonstreaming.py:41]: isn't device map Union[torch.device, dict, str]? Think this is wrong

**Review by marksverdhai — COMMENTED** [review]
Nice work on the streaming primitives, @gerayking. The model-level token streaming iterator and async decode pipeline are well structured. A few things I wanted to flag from our experience running a similar streaming setup in production:

## Scheduler integration

The biggest gap I see is that `forward_streaming()` returns `OmniOutput` with `"finished": torch.tensor(is_finished)`, but the generation scheduler (`omni_generation_scheduler.py`) doesn't know to look for this field. Without a corresponding scheduler change, the request will be marked `FINISHED_STOPPED` after the first `forward()` call, so subsequent chunks will never be generated in the online serving path.

In our fork we added an `is_final` flag to `OmniOutput.multimodal_outputs` and patched the scheduler to keep requests in `RUNNING` status until `is_final=True`. Worth considering whether this PR should include that plumbing, or if it's intentionally scoped to offline inference only for now.

## Serving layer

Related to the above — there are no changes to `serving_speech.py` or any entrypoint. For online serving you'd need a `StreamingResponse` path that yields chunked audio (PCM or WAV with a placeholder content-length header). Is this planned as a follow-up?

## Stride-0 tensor serialization

One bug we hit in production: when chunks cross certain boundaries, the audio tensor returned from the decoder can have stride-0 dimensions (from `expand()` or numpy views). Calling `.contiguous()` alone doesn't fix stride-0 — you need an explicit `.clone()`. This manifests as a crash during ZMQ serialization in the connector layer. Something to watch for when this gets wired into online serving.

## Context stripping accuracy

The `AsyncDecodingPipeline._decode_worker` uses a fixed `upsample_rate` (`getattr(..., "decode_upsample_rate", 2000)`) to convert context frames to context samples. In practice the transformer-based decoder doesn't always produce exactly `frames * upsample_rate` samples, which can cause subtle audio glitches at chunk boundaries. A proportional calculation (`context_frames / total_decoded_frames * total_decoded_samples`) adapts to the actual decoder output and is more robust.

## Streaming state cleanup

In `forward_streaming()`, the streaming state is only cleaned up on a *subsequent* call after `is_finished=True`. If the caller stops iterating once it sees `finished=True` (as the docstring suggests), the generator and accumulated audio chunks will leak in `self._streaming_state`. Cleaning up immediately when the generator signals completion would prevent this.

## Minor

- `print(f"chunk_size : {chunk_size}, {runtime_additional_information}")` on line 142 of `qwen3_tts.py` looks like a debug print that should be removed or converted to `logger.debug()`.
- The repetition penalty loop iterates per unique token ID with Python-level indexing. A vectorized approach using `torch.where()` on the unique ID set would be faster for longer sequences.

The decoder-side fixes in this PR (`codec_valid_max` truncation in `chunked_decode`, `use_cache=False` in the tokenizer, `suppress_tokens` on the code predictor) are solid and independently valuable regardless of the streaming work.

Happy to help with the scheduler/serving integration if that's useful.

**Comment by gerayking** [comment]
> Nice work on the streaming primitives, @gerayking. The model-level token streaming iterator and async decode pipeline are well structured. We've been running a streaming TTS setup in production on [our fork](https://github.com/heiervang-technologies/ht-vllm-omni/tree/ht), so I wanted to share some context on how our approaches compare and flag a few things.
> 
> ## How our fork approaches streaming vs this PR
> We took a slightly different architecture. Rather than a threaded `AsyncDecodingPipeline`, we do synchronous inline decoding — the codec-to-audio decode happens in the same thread as the generator, without queues. The tradeoff is less potential parallelism between token generation and audio decoding, but simpler concurrency and no queue-related edge cases (like the bounded `stop()` put).
> 
> At the token generation level, we have a single `generate_streaming()` generator on the `ForConditionalGeneration` class rather than a separate talker-level iterator. We also factored sampling and repetition penalty into standalone helper functions (`_sample_token()`, `_apply_repetition_penalty()` using vectorized `torch.where`) instead of inlining them in the loop.
> 
> For context stripping when decoding chunks, we use a proportional ratio (`context_frames / total_decoded_frames * total_decoded_samples`) rather than a fixed `upsample_rate`. The transformer-based decoder doesn't always produce exactly `frames * upsample_rate` samples, so the proportional approach adapts to the actual decoder output and avoids subtle audio glitches at chunk boundaries.
> 
> The bigger architectural difference is in the integration layers — which brings me to the main gaps I see in this PR:
> 
> ## Scheduler integration
> `forward_streaming()` returns `OmniOutput` with `"finished": torch.tensor(is_finished)`, but the generation scheduler (`omni_generation_scheduler.py`) doesn't know to look for this field. Without a corresponding scheduler change, the request will be marked `FINISHED_STOPPED` after the first `forward()` call, so subsequent chunks will never be generated in the online serving path.
> 
> In our fork we added an `is_final` flag to `OmniOutput.multimodal_outputs` and patched the scheduler to keep requests in `RUNNING` status until `is_final=True`. The scheduler reads `is_final` from the pooler output and only transitions to `FINISHED_STOPPED` when it's true. This required handling deserialization edge cases too — `is_final` can arrive as a bool, list, tensor, or scalar after ZMQ transport.
> 
> Is this PR intentionally scoped to offline inference only, or is the scheduler/serving integration planned as a follow-up?
> 
> ## Serving layer
> Related — there are no changes to `serving_speech.py` or any entrypoint. For online serving you'd need a `StreamingResponse` path that yields chunked audio. In our fork we added `_stream_progressive_audio()` as a FastAPI `StreamingResponse` that yields PCM bytes or WAV (with a placeholder max-size WAV header). One non-obvious thing: the output processor accumulates all tensors across iterations (not just new ones), so you need a cursor to track which chunks have already been yielded.
> 
> ## Stride-0 tensor serialization
> One production bug we hit: when chunks cross certain boundaries, the audio tensor from the decoder can have stride-0 dimensions (from `expand()` or numpy views). Calling `.contiguous()` alone doesn't fix stride-0 — you need an explicit `.clone()`. This crashes during ZMQ serialization in the connector layer. Something to watch for when wiring this into online serving.
> 
> ## Streaming state cleanup
> In `forward_streaming()`, the streaming state is only cleaned up on a _subsequent_ call after `is_finished=True`. If the caller stops iterating once it sees `finished=True` (as the docstring suggests), the generator and accumulated audio chunks will leak in `self._streaming_state`. Cleaning up immediately when the generator signals completion would prevent this.
> 
> ## Minor
> * `print(f"chunk_size : {chunk_size}, {runtime_additional_information}")` on line 142 of `qwen3_tts.py` looks like a debug print that should be removed or converted to `logger.debug()`.
> 
> ## What this PR does well
> The decoder-side fixes in this PR are solid and independently valuable:
> 
> * `codec_valid_max` truncation in `chunked_decode` — prevents out-of-bounds embedding lookups
> * `use_cache=False` in the tokenizer 12Hz decoder — prevents KV cache leaks
> * `suppress_tokens` on the code predictor — prevents generating tokens >= 2048 that would crash the decoder
> 
> These are fixes we don't have in our fork yet and plan to adopt.

Thanks for the detailed review and sharing your production experience!

Scope clarification: This PR is intentionally scoped to offline inference only. Scheduler integration and online serving changes are planned as follow-up work.

Will fix in this PR:
- Remove debug print() on line 142
- Add .clone().contiguous() for stride-0 tensor safety
- Clean up streaming state immediately on completion

**Comment by X-Skoprio** [comment]
Hi, is this going to be online anytime soon ? Thank you !

File: vllm_omni/model_executor/models/qwen3_tts/qwen3_tts.py (hunk 1 above)
Comment (gerayking) [reply vllm_omni/model_executor/models/qwen3_tts/qwen3_tts.py:184]: Externally we only expose forward. If a model needs to support streaming output, it should additionally implement forward_streaming; otherwise forward alone is sufficient.

**Comment by gerayking** [comment]
> Hi, is this going to be online anytime soon ? Thank you !

Qwen3-TTS online serving streaming output will be supported in another PR. I’ll push it soon.

**Comment by hsliuustc0106** [comment]
I think there are some bugs in this PR, we should not have changed so many files :)

**Comment by hsliuustc0106** [comment]
## 🤖 Code Review: PR #1189 — Streaming Audio Output 🔴

**Verdict**: Request Changes

**Summary**: Adds streaming audio output for TTS. However, the PR is ~132KB and bundles unrelated changes: CI infrastructure (pytest markers, nightly tests), ComfyUI integration app, GitHub workflows, and CFG parallel mixin docs. The PR must be split before meaningful review of the streaming core is possible. The CI changes and ComfyUI integration are entirely separate concerns that increase merge conflict risk and make bisecting regressions impossible.

**Key Concerns:**
- ⚠️ PR must be split into at least 3 PRs: (1) Streaming TTS core, (2) CI infrastructure (pytest markers, nightly config, buildkite changes), (3) ComfyUI integration (new apps/ComfyUI-vLLM-Omni/ directory). Bundling makes review difficult and bisecting regressions impossible.
- ⚠️ Cannot fully assess the streaming implementation due to diff size. Key unanswered questions: How are audio chunks delivered to the client? Does it integrate with existing chunked_decode() on the 12Hz tokenizer? What is the chunk boundary strategy?
- ⚠️ Interaction with PR #1205 (CUDA Graph) needs clarification -- both modify the decode path and could conflict. Streaming chunk size must align with CUDA graph capture sizes.
- ⚠️ First chunk latency measurement against the 200ms target cannot be verified without isolating the streaming core.

### Detailed Feedback

#### 📄 `.buildkite/pipeline.yml`

🟡 **MEDIUM** (architecture)
> CI pipeline changes (switching from explicit test paths to pytest marker-based selection with '-m core_model and cpu') are a good pattern but are completely unrelated to streaming audio. Should be in a separate PR.

#### 📄 `.buildkite/test-nightly.yaml`

🟡 **MEDIUM** (architecture)
> New nightly test config for H100 is reasonable but unrelated to streaming audio. Should be in the CI infrastructure PR.

#### 📄 `.github/workflows/build_wheel.yml`

🟡 **MEDIUM** (architecture)
> Workflow modifications are unrelated to streaming audio output. Should be in the CI infrastructure PR.

#### 📄 `apps/ComfyUI-vLLM-Omni/`

🟠 **HIGH** (architecture)
> Entirely new ComfyUI integration application. This is a separate feature that should not be bundled in a streaming audio PR. Creates unnecessary merge conflicts and makes the PR unreviewable at 132KB.

#### 📄 `vllm_omni/entrypoints/openai/serving_speech.py`

🟠 **HIGH** (architecture)
> Cannot fully assess streaming integration with create_speech() due to the PR's size. Key questions: (1) How are audio chunks delivered -- chunked HTTP transfer, SSE, or WebSocket? (2) Does it integrate with the existing chunked_decode() on the 12Hz tokenizer? (3) What is the chunk boundary strategy and does it match CUDA graph PR #1205 assumptions?

### Suggestions

1. Split the PR immediately into 3 separate PRs: streaming core, CI infrastructure, ComfyUI integration.
2. Ensure the streaming chunk size aligns with CUDA graph capture sizes from PR #1205.
3. Add a benchmark measuring first-chunk latency specifically (PR #1408 notes this metric requires streaming).
4. Clarify the audio chunk delivery mechanism (chunked HTTP, SSE, WebSocket).
5. After splitting, rebase the streaming core on top of PR #1205 to avoid decode path conflicts.


*Part of [Issue #938](../issues/938) — Qwen3-TTS Production Ready*

---
*Review generated by vLLM-Omni PR Review Agent*

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by lishunyang12** [comment]
The size of PR make it difficult to review and test, can we downsize the scope? 

Edit: most of changes are not from this PR.

**Comment by lishunyang12** [comment]
Looks like the branch was forked from an older main and carries ~28 already-merged commits (CI, ComfyUI, docs, etc.) — that's where the 149 files come from. The actual streaming work is only ~6 files.

`git rebase origin/main` should clean this up. If that gets messy, cherry-picking your streaming commits onto a fresh branch from main works too.

**Comment by gerayking** [comment]
> Looks like the branch was forked from an older main and carries ~28 already-merged commits (CI, ComfyUI, docs, etc.) — that's where the 149 files come from. The actual streaming work is only ~6 files.
> 
> `git rebase origin/main` should clean this up. If that gets messy, cherry-picking your streaming commits onto a fresh branch from main works too.

sry, I will fix it tonight.

**Comment by hsliuustc0106** [comment]
@vllm-omni-reviewer

**Comment by hsliuustc0106** [comment]
check #1438 for details
